### ISSUE TITLE
No tracking of other metrics than loss/perplexity

### ISSUE THREAD

**Issue by serhiy-yevtushenko** [bentrevett/pytorch-seq2seq#21]
This one is a more conceptual issue: when developing seq2seq model (or neural network models in general), one should not rely only on loss to judge the quality of the model (unless it's a regression task).

In these examples, the usage of more appropriate metrics during the evaluation (for example, accuracy, if aiming for exact translation, BLEU score or Levenstein distance from the target) will be more illustrative for me and would show, whether the actual learning occured.

**Comment by bentrevett** [comment]
The issue is that the models used in these tutorials use considerably less parameters and are trained for less time than models used in production and in the papers they are based off. This would cause the BLEU scores from the tutorials to look pretty bad in comparison. 

I also feel like a value of loss is something that's easier to understand (lower is better) than a BLEU score. Each model uses the same loss function so we can directly compare them in terms of their losses. If I added BLEU score, then we'd just be looking for higher BLEU score. A lower loss generally (although admitted not always) gives a higher BLEU score. So why bother adding the BLEU when we can just look at the loss? 

Saying all of that, I am still considering adding the BLEU score as it would be useful to show - this is supposed to be a tutorial after all - but would like to find a clean implementation that doesn't use any libraries outside of PyTorch and TorchText.

**Comment by serhiy-yevtushenko** [comment]
I would suggest using some kind of smaller/generated dataset in order to show the comparison of the models. (Possible suggestion - some kind of character reversal task, converting between different format of numbers (decimal based - binary based) as a string or some other.

**Comment by bentrevett** [comment]
I like that idea, but if the toy task is too trivial then all of the networks will be able to solve it and it will not show any useful way of comparing between models. 

I'm also not too confident with making the statement that "model X is better than model Y on a toy sequence task therefore model X is better than model Y at real world sequence tasks". 

I'll try and think of some ideas.

**Comment by serhiy-yevtushenko** [comment]
Concerning the clean implementation of bleu - here is one that seems suitable (although I would say, that it make sense to test it first)

https://github.com/MaximumEntropy/Seq2Seq-PyTorch/blob/master/evaluate.py

**Comment by bentrevett** [comment]
Thanks for sharing. I've not been able to get much GPU access time recently, but when I do I'll give this a go.

**Comment by serhiy-yevtushenko** [comment]
For GPU access one could try https://colab.research.google.com/. Just create new notebook (or upload existing one), and change runtime type  (Runtime-> Change Runtime Type) to GPU

**Comment by bentrevett** [comment]
TorchText is adding metrics in the next update, as such I've now added BLEU metrics (using their implementation) to notebook 4 and will add them for the notebooks that follow.

Only took me ~6 months!
